### ISSUE TITLE
Add ability to limit auto-detection to a subset of languages

### ISSUE THREAD

**Issue by sindresorhus** [ggerganov/whisper.cpp#1242]
From what I can tell, auto-detection simply picks the language with the highest probability score. Sometimes, I know the language could only be 1 out of 5 possible languages. In such cases, it would be useful to be able to specify the possible languages, to improve the likelyhood of the auto-detection picking the correct language.

**Comment by ggerganov** [comment]
You can use the `whisper_pcm_to_mel()` + `whisper_lang_auto_detect()` API.
You will get the probs for all languages in the `lang_probs` array:

https://github.com/ggerganov/whisper.cpp/blob/59a3d0cb576db605f76f82f07350647837e15c7a/whisper.h#L244-L255

**Comment by sindresorhus** [comment]
Thanks for the hint. That definitely works. It would still be nice to have a single param for this though.

**Comment by misutoneko** [comment]
[whisper.cpp_limit_language_autodetection_patch.diff.gz](https://github.com/ggerganov/whisper.cpp/files/12539830/whisper.cpp_limit_language_autodetection_patch.diff.gz)

Here's a little patch you can try.  This will extend the "auto" parameter in the main example so that you can give it a list of allowed languages.
So instead of -l auto you would use something like -l auto:pt,es,sv,en

Please note that although this *seems* to be working, I won't be making a PR out of it. But feel free to use the code as you wish.

**Comment by JayPatel095** [comment]
Running into the same issue above. The fix is small, we can filter while building logits_id instead of after in `whisper.cpp`: 

```
for (const auto & kv : g_lang) {
    const auto lang_id = kv.second.first;
    if (!allowed(lang_id)) continue;                  
    # allowed() checks lang_id against the passed in --allowed-languages [l1, l2...] if it exists

    const auto token_lang = whisper_token_lang(ctx, lang_id);
    logits_id.emplace_back(state->logits[token_lang], lang_id);
}
```
Softmax renormalizes over just the subset this way.

misutoneko's patch filters inside the sort comparator, which mutates elements mid-sort, and assumes 2-char lang codes which won't work with newer haw and yue. I'd add `allowed_lang_ids` + `count_allowed_lang_ids `to whisper_full_params (same idea as `grammar_rules` and `n_grammar_rules`) and an --allowed-languages flag on the CLI. 

@ggerganov What do you think? Seems like a relatively common issue. Let me know if that works and I'll open a PR.
